opentelemetry/proto/metrics/v1/metrics.proto

```diff
@@ -227,9 +227,39 @@ message MetricDescriptor {
   // temporality is the Temporality of values this metric has.
   Temporality temporality = 5;
 
+  // Monotonic describes the relationship between successively reported
+  // values of a metric (i.e. values of a metric with the same labels are
+  // non-decreasing).
+  //
+  // This refinement of metric values can be useful in understanding how to
+  // deal with discontinuities in the data (i.e. calculating rates of the
+  // data without introducing artifacts from a reset).
+  //
+  // This refinement is not coupled to any of the other metric qualities.
+  // For example, a CUMULATIVE metric is not implicitly NONDECREASING. This
+  // decoupling is needed in order for this protocol to agonistically
+  // transport metrics. However, it is important to note this diverges from
+  // some other metric systems (e.g Prometheus).
+  enum Monotonic {
+    // UNSPECIFIED_MONOTONIC is the default, and means the monotonic nature
+    // of the metric values is unknown.
+    UNSPECIFIED_MONOTONIC = 0;
+
+    // NONDECREASING means all the successive metric values increase or
+    // remain constant.
+    NONDECREASING = 1;
+
+    // NONINCREASING is currently omitted due to the lack of a known
+    // use-case. If you have a fitting use-case please open an issue to add
+    // this Monotonic value.
+  }
+
+  // monotonic describes the Monotonic refinement of values this metric has.
+  Monotonic monotonic = 6;
```

Comment:
Based on a comment from gitter:
> I think we can have only `Domain`, and we don't need "monotonicity". For example if domain refers to the raw data not to the aggregated data, then by having a non negative values domain implies nondecreasing monotonicity. The advantage being that we can apply this to data in "distributions" and not only to int64/double.

Currently the "monotonic" property is hard to correlate with Histogram/Summary (or even the newly proposed Distribution in that closed PR).

Reply:
The problem with only including `Domain` and not `Monotonic` is that `Domain` does not imply `Monotonic`. For example, if data for a metric continuously switches back-and-forth between say `0` and `1`, it is in the `Domain` of "greater than or equal to zero", but it is not monotonic. If you calculate the derivative on this data to generate its rate thinking it is monotonic it will result in large artifacts due to the discontinuous nature.

Conversely, it is also true that `Monotonic` does not imply `Domain`. So I don't think it is a choice of overlapping concepts to be made here.

I agree that the monotonic nature of Histogram/Summary data is not intuitively defined. But I don't think you improve the situation by not including this information. Take for instance when you want to translate into Prometheus metrics, without including a description of the monotonic nature of the data you would not be able to translate into any data into Prometheus Histograms or Counters.

I think instead, I can add better descriptions of how this monotonic refinement relates to the exposition formats. How does that sound?

Reply:
I do agree that if we cannot make sure that it implies "monotonicity" we should not apply that proposal.

Based on this example I think I see something wrong in my thinking, for me int64/double types meant SumInt64/SumDouble, so you cannot get 0 -> 1 -> 0 unless you have a "-1" somewhere:
> For example, if data for a metric continuously switches back-and-forth between say 0 and 1, it is in the Domain of "greater than or equal to zero", but it is not monotonic. If you calculate the derivative on this data to generate its rate thinking it is monotonic it will result in large artifacts due to the discontinuous nature.

Reply:
Is that the difference? You think for int64/double that they may not be a sum?

Reply:
Just a random thought: Maybe another way to think about this is domain + temporality determines monotonicity, in your example of 0->1 transitions that will be an instantaneous metric, so will always be a gauge in Prometheus. If temporality != Instantaneous then we can determine the monotonicity?

Reply:
Yes, I think that is the difference.

Reply:
What do you think about my latest comment (https://github.com/open-telemetry/opentelemetry-proto/pull/145#discussion_r417000445), is that a correct assumption?

Reply:
> Just a random thought: Maybe another way to think about this is domain + temporality determines monotonicity, in your example of 0->1 transitions that will be an instantaneous metric, so will always be a gauge in Prometheus. If temporality != Instantaneous then we can determine the monotonicity?

Yeah, I'm currently thinking this exact same thing through. I worry there might be a looseness with terms though.

If the metric is CUMULATIVE is there really a guarantee that the data is monotonic? Based on the [new UpDownSumObserver](https://github.com/open-telemetry/oteps/pull/98/files?short_path=95b3f96#diff-95b3f96b8945175022788618354f953e) I think that is going to be possible.

Reply:
After I wrote that random thought I now have the impression that monotonicity does not apply to any instantaneous metric, correct? :)) Probably I need to go have a drink or sleep :)))

Reply:
> If the metric is CUMULATIVE is there really a guarantee that the data is monotonic? Based on the new UpDownSumObserver I think there is going to be a situation where that is not the case.

No, what I am trying to say is: if the metric is cumulative or delta and domain is not-negative then is monotonic. if the metric is instantaneous is not monotonic.

Reply:
> Probably I need to go have a drink or sleep :)

Same. :smiley: 

> ... if the metric is cumulative or delta and domain is not-negative then is monotonic.

I think that is true if domain describes the domain of the measurement an instrument can make. The way I'm using these refinements is to describe the quality of the aggregation values.

It is true that if you only ever put non-negative values through a sum aggregator the resulting aggregation is monotonic. However, if you instead say that the resulting aggregation values are non-negative you cannot say anything about the monotonicity of the that aggregation.

Are referring to different parts of the pipeline when we are using these terms to describe data?

Reply:
Yes I think I had that in the initial comment "For example if domain refers to the raw data not to the aggregated data". That was the idea :)

Reply:
Now that we agree on that, do you think actually referring to the raw data domain is better? Argument being that we can determine interesting properties for Distributions as well as for sum values, for example in a Distribution we can determine if the sum is monotonic (and that percentiles are positive and never going to have negative buckets, etc).

Alternative we can go with your approach and extend the comment that monotonic applies to: int64/double delta or cumulative and to sum in the distribution if delta or cumulative. I think monotonicity should not apply to instantaneous aggregations.

Reply:
> Now that we agree on that, do you think actually referring to the raw data domain is better? Argument being that we can determine interesting properties for Distributions as well as for sum values, for example in a Distribution we can determine if the sum member is monotonic (and that percentiles are positive and never going to have negative buckets, etc).
> 
> Alternative we can go with your approach and extend the comment that monotonic applies to: int64/double delta or cumulative and to sum in the distribution if delta or cumulative. I think monotonicity should not apply to instantaneous aggregations.

I'm not 100% sure what the best approach here is.

It seemed to me like describing the data contained here would be the most useful to the user of this proto instead of making them infer things from understanding the OpenTelemetry processing pipeline. Additionally, it means that changes in the OpenTelemetry instruments/processors/aggregators are agnostically supported here. This would match the Type property as that describes the way the data is stored here.

That being said, it seems like the Temporality quality is already describing how the raw data is combined in the OpenTelemetry pipeline, so there is precedence in both ways.

I'm interested what @jmacd and @c24t think about this. @c24t had raised a similar question on the initial #137.

Reply:
Resolving with the decision having been to describe the measurement value as it is kept in this proto (instead of how it was provided to the measuring instrument), and that we move in a new direction to have the monotonic refinement directly linked to the type.